### ISSUE TITLE
Github Workflows change proposal

### ISSUE THREAD

**Issue by Enrico2** [apollographql/rust#71]
Hi,

I spent a good amount of time yesterday thinking about our CI/CD process in this repository and realized we might want to do things a little differently. 

I had a few motivators, the merit of which may or may not be subjective :)
1. Make PR checks finish as quick as possible without significantly affecting the signal we get from them.
1. Fail fast if there's a build failure.
1. Heavily cache so that we can benefit from incremental compilation and avoid duplicate work.

The first thing I noticed was that for each PR commit, we're running the same build 6 times (3 for separate OS targets, one in build.yml and one in test.yml) and tests 3 times.
In rust, running `cargo test` also builds, and because these run in parallel workflows, we wait a long time before tests are actually run. This often means a long time before we get a good signal from the check.

Then I saw that we build a full release binary on every PR commit, something that I'm guessing we don't need 99% of the time.

I started to play around with different variations, these were not attempts at changing the workflows, but rather exploration. You can see the different commits in [this draft](https://github.com/apollographql/rust/pull/70).

At some point, and this became a distraction, a unit test was consistently failing on something very unexpected; a _timeout_ from the wiremock server we spin up, I started digging into that, but because it was only reproducible on CI, my iteration speed was slow af because I noticed the cache is not doing what it's supposed to and we download and compile all of the cargo stuff on each build. The cache always misses. Upon further investigation, this is probably something I BROKE in this branch, but anyway, it got me to think about what we cache, and made me notice another thing, we build the cache key by hashing
```
${{ hashFiles('**/Cargo.lock') }}
```
But, we don't actually check in `Cargo.lock` so that part of the key is empty, which means we may change dependencies' versions but load older dependencies from cache. **This is real bug no. 1**

Along the way, I saw that while we build the entire repo on mac and windows, we only build the CLI on Linux. **This is real bug no. 2**

Back to the main thing, I asked @jbaxleyiii why we've set up the workflow to build a full release CLI on each PR commit, and he replied:

> On each PR we build a release quality build of the binary which collaborators can download from Github and use as a fully built binary to try out. We do this for Apollo Server (the ability to give out per PR builds to people) and it has been extremely helpful for the community and customers alike. Our setup here isn't quite as nice as it requires one of us to download the artifact for people but the basic idea is still there. This may be overkill, but @trevor-scheer and @abernix  can attest to the power of a durable per PR build of the project that you can give to customers to try out without going through any kind of release cycle. It also helps to spot release time issues at the PR time instead of when you want to cut the release and now have to fight through some build step issue you didn't expect.

That took me to another issue; We define "workflows" in files, each of which has jobs that can run in parallel, or as a DAG. We have control inside a workflow, but we can't control anything between workflows.
I noticed that we're not mapping our capital-W Workflows to our actual workflows.

Our defined workflows are:
```
audit.yml
build.yml
deploy-cdn.yml
release.yml
test.yml
```

Our actual day to day workflows are more like
- Rust PRs
- CDN PRs
- Deploy the CDN
- Build a CLI from a branch
- Release a CLI version update (changelog and all)
- Rust nightly security audit
- Merges to `main` (maybe?)

And so, what I'm proposing is that we match our workflows to our Workflows, and for each Workflow, optimize it to when it's used. My straw dog proposal is this:

## Rust PR Workflow
- **event**: `pull_request` on non `cdn/**` changes.
- Job 1: Lint and style.
	- Annotate with [Clippy warnings](https://github.com/actions-rs/clippy-check). This doesn't fail the job, just gives reviewers a hint that the warning is there.
	- Check formatting (this should fail, but not stop the rest of the jobs)
- Job 2: `Cargo test` on all OSs. When running `test`, it builds and tests the entire workspace, we should also test with verbose and debug output to make debugging easy. Specifically `RUST_BACKTRACE=1 RUST_LOG=debug cargo test -- --nocapture`
  If we can get coverage to work, we should, as a signal, not as a hard pass/fail requirement.

Note that building everything also builds a binary for all environments, we'll get signal if that fails in any way, the one thing we _don't_ need to on every PR commit, is to build a `--release` binary and publish the binaries. If we need them, we can still build them for a specific branch (see below)

## CDN PRs
- **event**: `pull_request` only on `cdn/**` changes.
- lint + build + test + coverage as above.

## Deploy the CDN - CD when merging to `main`
No changes on this one 👍 

## Build a `release` CLI from a branch.
- **events**: 
	- Merge to `main`  
	- `workflow_dispatch` ([see this new feature!](https://github.blog/changelog/2020-07-06-github-actions-manual-triggers-with-workflow_dispatch/))

This workflow is not something we need to do often. When we do need it, we only really need to build the CLI and its dependencies, but as the rust repository grows, that might not mean building the entire repo.
	  This runs `cargo build --release` for all 3 OSs, with their appropriate flags that we need to pass in, and saves the results as artifacts.

## Release a CLI version update
- **event**: tagging a commit with `v*`.
- test (this builds a `debug` target)
- `build --release` (this builds a `release` target)
- create binaries, SHAs etc.

The workflow jobs and steps here are **very** similar to the steps when releasing the CLI from a branch. We can use the same workflow that is triggered on multiple events, and the diff between workflows can branch on conditions on the event type.

## Rust nightly security audit
No changes on this one 👍 

## General stuff:
- We can significantly DRY up `release.yml` by putting more info in the matrix instead of manual environment variables and using a consistent path for releases.

- Use actions-rs' versions of installing the toolchain for a target and their version of running cargo, to get better updates, control, and action reports. Currently, we wrote these ourselves.

This proposal is implemented in https://github.com/apollographql/rust/pull/84

**Comment by Enrico2** [comment]
Relevant: https://github.com/apps/merge-when-green

**Comment by Enrico2** [comment]
Iterating on some ideas in https://github.com/apollographql/rust/compare/ran/workflow_revamp

**Comment by ashleygwilliams** [comment]
> Rust PR Workflow

I don't think it makes sense to lessen the builds here. Lint and style is useful, as would be testing on all platforms. CI for PR workflow is meant to ensure that we don't merge anything to main that doesn't pass tests. Therefore we should be testing all platforms, as I would not think it'd be acceptable to have a PR merged that broke Windows support.

-------

Stepping back: I would really like to hear more about the *problem* that we are solving with this effort. I am not convinced that the situation as we currently have it is particularly burdensome and would warrant such efforts. I think a lot of the reasonable changes we could make are tradeoffs we've caused due to using a monorepo, which I think will only scale so far before we'll need to split it, in which case I'm not convinced that a ton of effort now would be worthwhile.

------

Re: merge when green, I am generally in favor, however, when merging a lot of PRs at once, perhaps right before a release, it can be helpful to do them in an order, and the manual process can be helpful for that. I'm not opposed to the bot but I also am not super convinced that we have enough pain right now to justify automating it!

**Comment by Enrico2** [comment]
Replying a bit of order for better context:

> Stepping back: I would really like to hear more about the _problem_ that we are solving with this effort.

The problem I've experienced is around the time it takes to get signal from the checks on whether or not a commit is greed/red. It's currently not fast enough so as not to leave you hanging, and not slow enough to context switch and come back later. I've found that if I'm actively iterating (both when I'm the author or when I'm the reviewer), the delay has felt like hitting strong head wind.

I'll take the monorepo discussion to #82 

> I would not think it'd be acceptable to have a PR merged that broke Windows support.

Yeah, as I was iterating yesterday I went back on my position to limit the environments during iteration. I do think however that we can improve the parallelization there, and fix the caching (#70 WIP, I'll update the description there with the goals).

Ideally though, if there was a way to run a pre-merge check, that's not on `pull_request` (which would run on every commit), I'd still prefer that as it would give us exactly what we want; fast commit iteration, and full validation before merging. I couldn't see a native way to do this, and wiring it ourselves with some pre-flight branch or using label events for example did seem like overkill.

One thing though that we really don't need on every single commit, is building a full, release quality binary. I think we can have `cargo test` on all OSs on every commit, but move the various `cargo build --release` to a manual trigger that we use only when we need it (e.g. on `deployment`), this will be part of #70 

> when merging a lot of PRs at once, perhaps right before a release, it can be helpful to do them in an order

The plugin I linked above doesn't directly support Github actions easily, [this one does](https://github.com/pascalgn/automerge-action), both of them work according to what we define as "green" at the repo level, and we can require that a branch has to be up to date with main in order to be green, so effectively multiple PRs will merge one by one. We can configure this to also automate and bring a branch up to date when `main` updates.

-----------

To summarize, this problem may be subjective as to how much one might be bothered by slower checks. We should agree as a team what checks we want to have as a threshold to merging to `main`, while personally I may spend time to ensure that what we agree on is done in the most efficient way that's reasonable to implement.

**Comment by Enrico2** [comment]
I updated my proposal in the description.

**Comment by jbaxleyiii** [comment]
Following up from the [latest comment](https://github.com/apollographql/rust/pull/84#issuecomment-648851159).

I'm **really** interested in the idea of only building on merges to main. The builds do take a lot of time and we certainly don't have the need for the PR build workflow yet. I suspect that was a premature optimization on my part that is causing some early pain 😭. If we do this, how will it open an issue up?

As for the overall "what do we want CI/CD to look like" question:

Good OSS CI has a few properties that important to me to facilitiate effective collaboration and to protect us against breaking changes for a wide range of consumers. Which is basically what all CI is for but one thing I think is special about OSS is there is much lower threshold to "percieved pain" and "asked labor" we can ask of others. With GitHub's inline sugguestions, the ability to open a PR from GitHub, and the upcoming workspaces, not to mention when we move our docs infra over to this repo, I don't think we can have assumption that every commit to the repo will be backed with a local toolchain. *However*, the CI is not a local debugging tool either! CI, to me, is a wide gate for all contributors to the project. With that, I think the following is important:

- **Fast**: Slow CIs do leave me, personally, with a negative view of a project. This is probably unfair but it is definitely an base reaction that I have especially as I try to maintain and land multiple PRs in coordiation with each other. I don't use CI for debugging, but when there are issues that are isolate to a CI environment having a fast CI makes a big different. Moreso, having a fast RED/GREEN status helps me while I go from draft to readiness of a PR. My personal workflow is initial idea + local testing, open draft, verify CI passing, author detailed review guide for team them convert to full PR. A fast CI makes this more enjoyable for sure.

- **Consistent**: Flaky tests are the worst and they are much worse than a slow test. I think @Enrico2 has found this out for us with our lack of a stable cache key as we have had both slower than needed CI and odd failures at times with deps changing out from under us. Part of this consistnency, to me, is a plethora of environments running in parallel to narrow the chance or random errors cropping up. I orginally had nightlies turned on which I think we should turn back on to help guard against flaky tests being introduced

- **Quiet**: CI, again to me, is a gate / flag. A failure there is a marker for more local work to be done. I love the combination of a quite CI + exported xml reports with details that let me dive into locally to a problem to replicate it. One thing we are missing is a strong way to replicate a CI run locally with the cargo manifest and test output for faster debugging. I would prioritize fixing that over loud / spammy logs so that a solid 🔴 turns into a direct local debugging + fix. Otherwise my experience has been a lot of try / fail / CI spam that is so draining for me and for our CI budget 😅

Ulimately I think the problems that @Enrico2 is trying to solve is that our CI setup right now isn't an effective partner in his / our development flow. I do agree that there are some great improvements to be made here and a lot of the work in #84 look to do that. I'm curious to hear thoughts on the quiet / noisy bits as my view may very well be unique 😂.

All in all, I'm super excited to see a little time spent here speeding things up and moving to the community maintained build setup for the rust community instead of our home grown stuff. Thank you so much for kicking this off!

**Comment by Enrico2** [comment]
@jbaxleyiii I'd like to do something to move this along.

I think the only remaining bit is the verbosity of tests? I replied to that in the PR: https://github.com/apollographql/rust/pull/84#issuecomment-648548718
I'm also fine changing that if I'm the only one feeling there's value there.

I always gravitate towards any iterative improvement, so if we can at least classify the pieces in the proposal that are agreed upon for now, and we can make steps in that direction, and defer the rest?

**Comment by Enrico2** [comment]
I updated the proposal in what I hope is the last time, and I've updated https://github.com/apollographql/rust/pull/84 to match.

**Comment by Enrico2** [comment]
This was mostly relevant for a repo that had both the CLI and other stuff. Closing because the CLI is moving to its own repository.
